### ISSUE TITLE
Consider adopting semver.org versioning practices

### ISSUE THREAD

**Issue by stevenhaddox** [bbatsov/rubocop#4243]
## Expected behavior

I expected `bundle update` to grab the latest Rubocop that would still run properly in my gems for both a Ruby 1.9 and 2.x runtime.

## Actual behavior

My test suites broke in the Ruby 1.9 environments because Rubocop v `0.41.2` changed a core requirement in the version of Ruby needed to run Rubocop.

## Steps to reproduce the problem

I just had to hard-lock several gems to rubucop version `0.41.1` because version `0.41.2` required Ruby 2.0+ to be installed in order to run. For gems that want to support older versions of Ruby for longevity (and yes, I realize they've been EOL'd). This was an immediate issue upon running `bundle install`.

I had my Gemfile's locked pessimistically with `~> 0.41` to ensure I got the latest community standards for Rubocop, but doing so resulted in breaking changes to my gems. I'd love to see Rubocop adopt a semantic versioning system like that of semver.org so we can use a looser lock in the future version. If a version `1.0.0` had been cut for the change in a core requirement of Ruby (from 1.9 to 2.x) it would have saved me a few gem releases where I am hardcoding the older Rubocop in a patch release before I make a new major release for my gems that will now require Ruby 2.x or above.

I'm not expecting this to necessarily be adopted, but as it's commonly used in the community as a best practice and Rubocop is *the* gem to encourage best practices in our code, I thought I'd at least put in a request for consideration. Thanks for all the hard work you do!

## RuboCop version

Before: `< 0.41.2`
With the error: `0.48.1`

**Comment by Drenmi** [comment]
@stevenhaddox: I'm not sure you're arguing that it doesn't, but RuboCop does follow Semver:

> 4. Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

This is a deliberate choice, since things are still pretty volatile, and cutting 1.0.0 would introduce considerable friction in making changes.

It's a trade-off. It's harder for users to know exactly what they get, (reading the `CHANGELOG` is important) but it helps avoid a situation where development grinds to a halt because of having to tread to lightly.

(Personally I am violently opposed to supporting EOL versions, but that's a topic for another discussion. 😉)

**Comment by bbatsov** [comment]
@Drenmi is right. RuboCop's development is still in a state of flux - breaking changes are made every few releases and adopting SemVer right now would quickly push to major version to something like 20. Until we reach a state of relative stability in the core APIs and cops, we're going to stick to the current versioning scheme.

**Comment by ianfixes** [comment]
It sounds like you are aware of the problem (breaking changes are made every few releases) and the solution (adopt semver).  What does "a state of relative stability" look like, and what projects that you use have achieved it?

> adopting SemVer right now would quickly push to major version to something like 20

Are we in danger of running out of integers or something?

**Comment by kmk3** [comment]
(Link to a previous discussion for cross-reference: #5612)

**Comment by ianfixes** [comment]
If I'm understanding that previous discussion properly, the disagreement seems to boil down to just this:

> changes to cops will never result in major version changes, because they are not tied to RuboCop's API

I would argue that this is true in a purely technical sense, but not in a practical sense.  Put simply, the API is not the part of RuboCop that most people use.  In fact, the [documentation on installing RuboCop](https://docs.rubocop.org/rubocop/installation.html) explicitly suggests that the API _not_ be brought into one's project -- `require: false`

```ruby
gem 'rubocop', require: false
```

I submit that the "API" that users experience in practice is _solely_ the command line utility `bundle exec rubocop`.  The underlying API of the RuboCop executable is immaterial to users, provided that
* the same rubocop config file remains valid between versions
  * no changes to existing cop names (e.g. `Layout/RescueEnsureAlignment`)
  * no changes to parameter names within individual cops (e.g. `EnforcedStyleAlignWith`)
  * no removals or modifications of an option to a parameter within a cop (e.g. `start_of_line`)
* the new rubocop version will properly install against the same set of non-EOL ruby versions

**However**

I understand that there is also a crucial need to present and communicate a stable API to all the contributors who are writing the _cops_ for RuboCop.  As updates to the engine are made, all of the existing contributions need to adopt those changes -- a very significant effort distributed over a large number of contributors and requiring exhausting effort to properly test.

To resolve that, I wonder if RuboCop could hold a separate semver for its _rule-writing API_, in which each contributed rule could correspondingly record what semver of the rule engine it expects.  All the same busywork around announcing deprecations could still happen, but it would be presented _as warnings during `rspec` testing_ such that only contributors see them.  That would allow communication of the external functionality to the end user, but also communication of the internal functionality to the contributors. 

Is this a possible path forward?
